api/src/main/java/io/opentelemetry/common/Attributes.java

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.common;
+
+import static io.opentelemetry.internal.Utils.checkArgument;
+import static io.opentelemetry.internal.Utils.checkNotNull;
+
+import com.google.auto.value.AutoValue;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.TreeMap;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * An immutable container for attributes. The type parameter denotes the type of the values of the
+ * attributes.
+ */
+@Immutable
+public abstract class Attributes<T> {
```

Comment:
My concern with having it be named `*Map`, is that it doesn't implement the `Map` interface, and we don't really want it to. It might add more confusion to have a class that kind-of looks mappish, but doesn't implement the commonly known interface.

Reply:
If we make it package protected, though, I guess that's less of a concern. I wonder if there's another word than `Map` we can use, though, that describes the functionality more accurately.

Reply:
I do like the idea of the sub-classes, but since the creational patterns here are all based on static methods, we'll need to rethink how that works. Any thoughts on that? Do the subclasses need to contain the static `of` methods and a builder, as well? I'm playing around with ideas on this, but if you have some, please share. :)

Reply:
I'm going to open up a new PR with Labels, Attributes, and an ImmutableKeyValuePairs interface they both implement. I'll keep this open in case we think that other PR is garbage. See #1304